Hlib bondarev/workshop entity with address

HlibBondarev opened the pull request:
1) Updated Workshop entity - added new properties according to ISSUE - https://github.com/ita-social-projects/OoS-Backend/issues/1534.
2) Updated WorkshopBaseDto - added new properties (corresponding to new properties of the Workshop entity).
3) Updated WorkshopService methods Create() and CreateV2 for filling new properties of Workshop entity according to ISSUE - https://github.com/ita-social-projects/OoS-Backend/issues/1544..
4) Updated MappingProfile class for mapping new added properties of Workshop and Workshop dtos..
5) Updated TeacherConfiguration and WorkshopConfiguration classes to set new relations between entities.
6) Updated Create actions for V1-WorkshopController and V2-WorkshopController.
7) Added new NUnit tests for check new code.

HlibBondarev committed:
1) Removed unnecessary Using directives (using System and using Microsoft.EntityFrameworkCore) in WorkshopConfiguration class.
2) Removed unnecessary Using directive (using OutOfSchool.Common.Validators) in Workshop class.
3) Changed WorkshopBaseDto class.

HlibBondarev committed:
Fixed Workshop class.

HlibBondarev committed:
Changed the WorkshopConfiguration class.

HlibBondarev committed:
Changed the one-to-one relationship between Workshop and Teacher in the WorkshopConfiguration class.

HlibBondarev committed:
1) Removed the one-to-one relationship between Workshop and Teacher from the WorkshopConfiguration class.
2) Added the one-to-one relationship between Workshop and Teacher to the Teacher class.

HlibBondarev committed:
Added default value for CompetitiveSelection property in Workshop class.

HlibBondarev committed:
1) Renamed TeacherId property to DefaultTeacherId in WorkshopBaseDto.
2) Added DefaultTeacher property to WorkshopBaseDto.
3) Renamed TeacherId property to DefaultTeacherId in Workshop.
4) Fixed TeacherConfiguration class - HasForeignKey<Workshop>(x => x.DefaultTeacherId).

HlibBondarev committed:
Modified the Create action of the WorkshopController class: added code to populate the WorkshopDescription.Id property with a default value.

HlibBondarev committed:
1) Replaced IncludedStudyGroupTitles (type - List<string>) property with IncludedStudyGroups (type - List<WorkshopBaseDto>) in WorkshopBaseDto.
2) Changed MappingProfile - renamed IncludedStudyGroupTitles to IncludedStudyGroups.

HlibBondarev committed:
Added migration - UpdateWorkshopEntity.

HlibBondarev committed:
Added RequiredAttribute to AdditionalDescription property of Workshop class.

HlibBondarev committed:
Removed attribute - [Required(AllowEmptyStrings = true)] from Workshop class.

HlibBondarev committed:
Removed attribute - [Required(ErrorMessage = 'Additional description is required')] from AdditionalDescription property in Workshop class.

HlibBondarev committed:
Added migration - 20241005210750_UpdateWorkshopEntity.

HlibBondarev committed:
Added migration - 20241006190718_UpdateWorkshopEntity.

HlibBondarev committed:
Added migration - 20241006192707_UpdateWorkshopEntity.

HlibBondarev committed:
Added migration - 20241006193114_UpdateWorkshopEntity.

HlibBondarev committed:
1) Changed type IncludedStudyGroups property ( List<WorkshopBaseDto> replaced to ICollection<WorkshopBaseDto>) of WorkshopBaseDto class.
2) Added default value (new HashSet<WorkshopBaseDto>()) to IncludedStudyGroups property of WorkshopBaseDto class.

HlibBondarev committed:
1) Removed unnecessary Using directive (using OutOfSchool.Services.Enums) in Workshop class.
2) Added MemberOfWorkshop property to the WorkshopBaseDto class.
3) Modified the Create action of the WorkshopController class: added code to populate MemberOfWorkshop.Id and IncludedStudyGropes.Id with a default value.
4) Added mapping for MemberOfWorkshop and IncludedStudyGropes properties of Create method in WorkshopService class.
5) Changed WorkshopConfiguration and TeacherConfiguration classes: added DeleteBehavior.Restrict for link between entities - Workshop and Teacher.

HlibBondarev committed:
1) Replaced WorkshopId type with Guid? instead of Guid in Teacher class.
2) Changed TeacherConfiguration class - Ignore Workshop property.
3) Changed GetTeachersWorkshopId() method of TeacherService class, the last line of code -
'return teachersWorkshopId ?? Guid.Empty;' instead of 'return teachersWorkshopId;'.
4) Changed WorkshopConfiguration class:
added  One-to-One and One-to-Many relationships between Workshop and Teacher.
5) Added migration - 20241008150513_UpdateWorkshopEntity.

HlibBondarev committed:
The finished Create action with recursion

HlibBondarev committed:
Removed recursion code from Create action of WorkshopController.

HlibBondarev committed:
Added try-catch block to Create action of WorkshopController.

HlibBondarev committed:
Removed Newtonsoft.Json.PreserveReferencesHandling from Startup class.

HlibBondarev committed:
Fixed Create method of WorkshopService class.

HlibBondarev committed:
Fixed Create method of WorkshopService class.

HlibBondarev committed:
1) Fixed Create method of WorkshopService class - logger.LogInformation();
2) Fixed Startup class: added code line - .AddNewtonsoftJson().

HlibBondarev committed:
1) Changed WorkshopControllerTests class - added  teacherServiceMoq (type of Mock<ITeacherService>) and loggerMoq (type of Mock<ILogger<WorkshopController>>) fields;
2) Removed unnecessary using directive (using OutOfSchool.Services.Repository;) in WorkshopServiceTests class;
3) Removed unnecessary comment in SetIdToDefaultValue() method of WorkshopController class.

HlibBondarev committed:
Added tests for Create method of WorkshopService to WorkshopServiceTests class.

HlibBondarev committed:
Added rules for new properties of Workshop entity to Faker<Workshop> in WorkshopGenerator class.

HlibBondarev committed:
Renamed SetupCreateWithWorkshop(Workshop workshop, bool isMemberOfWorkshopIdExisted) method to SetupCreate(Workshop workshop, bool isMemberOfWorkshopIdExisted) in WorkshopServiceTests class.

HlibBondarev committed:
1) Fixed Exists() method of TeacherService class - removed interpolation in logging message templates.
2) Fixed SetIdToDefaultValue() method WorkshopController class. - replaced 'default' to 'Guid.Empty';
3) Changed Startup class - added code:
.AddNewtonsoftJson(options =>
{
    options.SerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore;
})
to handle self-references during serialization.
4) Added check to Create() method of WorkshopService class -
The main workshop for the workshop being created cannot be a member of another workshop.
5) Added appropriated test (see. p.4) to WorkshopServiceTests class.

HlibBondarev committed:
1) Added tests for Create() method of WorkshopServicesCombiner in WorkshopServicesCombinerTests class.
2) Added tests for Create() method of WorkshopServicesCombinerV2 in WorkshopServicesCombinerV2Tests class.

HlibBondarev committed:
Added tests for Create action of V2-WorkshopController to WorkshopControllerV2Tests class.

HlibBondarev committed:
Added 3 tests for Create action of V2-WorkshopController to WorkshopControllerV2Tests class.

HlibBondarev committed:
Added 3 tests for Create action of V1-WorkshopController to WorkshopControllerV2Tests class.

HlibBondarev committed:
Added a test for Create action of V1-WorkshopController to WorkshopControllerV2Tests class. Case - 'Is User Provider Owner Or Admin'.

HlibBondarev committed:
Fixed tests for Create action in V1-WorkshopController and V2-WorkshopController.

HlibBondarev committed:
Refactored code for 'Create Workshop' according to SonarCloud prompts.

HlibBondarev committed:
Simplified the SetIdsToDefaultValue() method using LINQ in WorkshopService class.

HlibBondarev committed:
Removed teacherServiceMoq fiield from WorkshopControllerTests and WorkshopControllerV2Tests classes.

HlibBondarev committed:
Removed migration - 20240925152617_AddWorkshopBusinessEntity

HlibBondarev committed:
Removed redundant comments in Workshop class.

HlibBondarev committed:
Resolve conflicts in Workshop class

HlibBondarev committed:
Resolved conflicts in WorkshopBaseDto and MappingProfile classes

HlibBondarev committed:
Resolved conflicts in WorkshopService and WorkshopController classes

HlibBondarev committed:
Resolved conflicts in WorkshopService and MappingProfile classes

HlibBondarev committed:
Resolved conflicts in WorkshopController

HlibBondarev committed:
Resolved conflicts in WorkshopService and MappingProfile classes

HlibBondarev committed:
Resolved conflicts in WorkshopController classes

HlibBondarev committed:
Resolved conflicts in WorkshopService

HlibBondarev committed:
Resolved conflicts in MappingProfile classe

HlibBondarev committed:
Resolved conflicts in WorkshopServiceTests

HlibBondarev committed:
Resolved conflicts in WorkshopService

HlibBondarev committed:
Resolved conflicts in WorkshopService, V1-WorkshopController, V2-WorkshopController and WorkshopServiceTests classes

HlibBondarev committed:
Resolved conflicts in WorkshopServiceTests class

HlibBondarev committed:
Fixing after merge-base develop

HlibBondarev committed:
1) Commented out code with Tags in Create() method of WorkshopService class.
2) Added and rolled out migration - 20241026101653_CreateWorkshop.

HlibBondarev committed:
1) Fixed adding Tags in Create () method of WorkshopService class.
2) Fixed test for WorkshopService in classes - SensitiveWorkshopsServiceDBTests, SensitiveWorkshopsServiceTests, WorkshopServiceDBTests and WorkshopServiceTests.

HlibBondarev committed:
1) Added attribute  [ModelBinder(BinderType = typeof(JsonModelBinder))] to TagIds property of WorkshopCreateUpdateDto class.
2) Changed MappingProfile class - deleted unnecessary comments.
3) Deleted mapping in the returned value of Create () method of WorkshopServicesCombiner class.
4) Removed this useless assignment to local variable 'id' - dto.TagIds?.ForEach(id => id = default);.

HlibBondarev committed:
1) Swapped the methods of Exists() and UpdateTeacher() in TeacherService class.
2) Removed unnecessary spaces in the code line of the PredicateBuild() method in WorkshopServices class.

HlibBondarev committed:
Cnaged SensitiveWorkshopsServiceTests class - swapped code lines in FetchByFilterForAdmins_FilteredBySearchString_ShouldReturnEntities() method.

AndrD1406 reviewed (approved):
Look well!
At least everything related to the tags entity

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -55,21 +55,25 @@ public class WorkshopBaseDto : IValidatableObject
+    [CollectionNotEmpty(ErrorMessage = "At least one DateTime range is required")]
     public List<DateTimeRangeDto> DateTimeRanges { get; set; }
 
+    [Required]
```
does this attribute make sense for non nullable fields ?

HlibBondarev replied:
O.k. 
Removed [Required] attribute from the bool type properties - IsPaid, IsInclusive, IsSelfFinanced and IsSpecial of WorkshopBaseDto class.

HlibBondarev replied:
WorkshopBaseDto class already uses data annotation attributes (like [Required], [MinLength], [MaxLength], [Range]) etc. Also we use checking  ModelState.IsValid property in Create() action of WorkshopControllerso so  implementing IValidatableObject provides a single, unified validation mechanism.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -125,6 +126,64 @@ public class WorkshopBaseDto : IValidatableObject
+    [ModelBinder(BinderType = typeof(JsonModelBinder))]
+    public AddressDto Address { get; set; }
+
     public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
```
I think it would be better to have it like extension method but not a part of the DTO

HlibBondarev replied:
WorkshopBaseDto class already uses data annotation attributes (like [Required], [MinLength], [MaxLength], [Range]) etc. Also we use checking ModelState.IsValid property in Create() action of WorkshopControllerso so implementing IValidatableObject provides a single, unified validation mechanism.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -125,6 +126,64 @@ public class WorkshopBaseDto : IValidatableObject
+    [Required]
+    [ModelBinder(BinderType = typeof(JsonModelBinder))]
+    public AddressDto Address { get; set; }
+
```
do we really need that DTO as a whole ?
I mean, 185 lines ? seriously ?
Can we split that to parts ?

HlibBondarev replied:
I'm thinking of splitting Workshop BaseDto into parts when I work on implementing draft saving in the backend. Then I plan to use a sequence of model inheritance.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopDto.cs:
```diff
@@ -8,7 +8,7 @@
-public class WorkshopDto : WorkshopBaseDto, IHasRating
+public class WorkshopDto : WorkshopCreateUpdateDto, IHasRating
 {
     public uint TakenSeats { get; set; } = 0;
```
why do we need this trailing `0`  ? Is it not a default value for `uint` ?
Why do we need default values in models at all ?

HlibBondarev replied:
O.k.
Removed default value for TakenSeats property of WorkshopDto.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ITeacherService.cs:
```diff
@@ -49,4 +49,11 @@ public interface ITeacherService
+    /// </summary>
+    /// <param name="id">Key in the table.</param>
+    /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.
+    Task<bool> Exists(Guid id);
```
`ExistsAsync` ?

HlibBondarev replied:
O.k. 
Reworked Exists() method to async method => ExistsAsync() in ITeacherService interface and its implementation in TeacherService class.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/TeacherService.cs:
```diff
@@ -176,6 +163,27 @@ public async Task<Guid> GetTeachersWorkshopId(Guid teacherId)
+        }
+        catch (DbUpdateException ex)
+        {
+            logger.LogError(ex, "Unreal to update teacher.");
```
Unreal -> Unable

HlibBondarev replied:
O.k.
Fixed Unreal -> Unable in TeacherService class.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/TeacherService.cs:
```diff
@@ -176,6 +163,27 @@ public async Task<Guid> GetTeachersWorkshopId(Guid teacherId)
+    /// <inheritdoc/>
+    public Task<bool> Exists(Guid id)
+    {
+        logger.LogInformation("Checking if Teacher exists by Id started. Looking Id = {id}.", id);
```
this is Trace or Debug but not Info message

HlibBondarev replied:
O.k.
Changed - used logger.LogDebug istead of logger.LogInformation in Exists() method of TeacherService class.

h4wk13 commented on OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/WorkshopConfiguration.cs:
```diff
@@ -13,25 +11,35 @@ public override void Configure(EntityTypeBuilder<Workshop> builder)
+            .WithOne()
+            .HasForeignKey(x => x.WorkshopId) // Shadow property for One-to-Many relationship
+            .IsRequired(false) // Optional relationship
+            .OnDelete(DeleteBehavior.Cascade);
```
Is it possible that teacher can be assigned to more than one workshop ?

HlibBondarev replied:
In this version - for each Workshop is creating new List of new Teachers.

h4wk13 commented on OutOfSchool/OutOfSchool.DataAccess/Models/Workshop.cs:
```diff
@@ -55,60 +31,113 @@ public class Workshop : BusinessEntity, IImageDependentEntity<Workshop>, IHasEnt
 
-    [Required]
-    public PayRateType PayRate { get; set; }
+    [MaxLength(500)]
```
I think it should be a const

HlibBondarev replied:
O.k. 
Create new constants - MaxAdditionalDescriptionLength, MaxPreferentialTermsOfParticipationLength and MaxKeywordsLength in Constants class. Used these constants for [MaxLength] attributes in Workshop class.

HlibBondarev committed:
1) Removed [Required] attribute from the bool type properties - IsPaid, IsInclusive, IsSelfFinanced and IsSpecial of WorkshopBaseDto class.
2) Removed default value for TakenSeats property of WorkshopDto.
3) Reworked Exists() method to async method => ExistsAsync() in ITeacherService interface and its implementation in TeacherService class. Used ExistsAsync() in private SetIdsToDefaultValue() method of WorkshopService
4) Fixed Unreal -> Unable in TeacherService class.
5) Changed - used logger.LogDebug istead of logger.LogInformation in Exists() method of TeacherService class.
6) Create new constants - MaxAdditionalDescriptionLength, MaxPreferentialTermsOfParticipationLength and MaxKeywordsLength in Constants class. Used these constants for [MaxLength] attributes in Workshop class.

HlibBondarev committed:
Added migration - 20241030104634_CreateWorkshop

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/WorkshopBaseDto.cs:
```diff
@@ -125,6 +125,61 @@ public class WorkshopBaseDto : IValidatableObject
+
+    public bool IsSpecial { get; set; } = false;
+
+    public uint SpecialNeedsId { get; set; } = uint.MinValue;
```
why `uint.MinValue` is better than nothing (which is `0`) or `0` or `default` ?

HlibBondarev replied:
Ok, replaced 'uint.MinValue' with '0' for uint properties of WorkshopBaseDto class.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -157,23 +138,17 @@ public async Task<WorkshopResultDto> CreateV2(WorkshopV2Dto dto)
         logger.LogInformation("Workshop creating was started.");
 
+        // TODO: after refactoring the DTOs for the Workshop entities, this method needs to be replaced with the correct mapping
+        await SetIdsToDefaultValue(dto); // This method sets the properties with the Id to the default value.
```
why do we need that ?

HlibBondarev replied:
We get the WorkshopCreateUpdateDto object from Frontend. It may contain not only null value but also invalid values, so we will have to replace them with the default value in the SetIdsToDefaultValue() method.
To avoid this, we can do it using mapping.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -1165,4 +1140,77 @@ await UpdateStatus(new()
+
+    private async Task<Workshop> CheckDtoAndPrepareCreatedWorkshop(WorkshopCreateUpdateDto dto)
+    {
+        if (dto.MemberOfWorkshopId.HasValue && !await Exists((Guid)dto.MemberOfWorkshopId).ConfigureAwait(false))
```
`why do we need that (Guid)` cast here ?

HlibBondarev replied:
MemberOfWorkshopId property of WorkshopBaseDto class has to be 'Guid?' type. Because of created Workshop may not be a member of other Workshop.
The Exists(Guid id) method has an argument of type 'Guid', not 'Guid?'. I don't want to overload or change this method, so I cast 'Guid?' to 'Guid'.
It's cast safely because I check dto.MemberOfWorkshopId.HasValue before doing so.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -1165,4 +1140,77 @@ await UpdateStatus(new()
+            createdWorkshop.Teachers = dto.Teachers.Select(mapper.Map<Teacher>).ToList();
+        }
+
+        createdWorkshop.Tags = (await tagRepository.GetAll()).Where(tag => dto.TagIds.Contains(tag.Id)).ToList();
```
this request looks strange.
So we are getting all tags and then filters them in memory ?
Can we add Where method to the repository instead ?

HlibBondarev replied:
Ok, I used tagRepository.GetByFilterNoTracking(tag => dto.TagIds.Contains(tag.Id)) method instead of (await tagRepository.GetAll()).Where(tag => dto.TagIds.Contains(tag.Id)).

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -1165,4 +1140,77 @@ await UpdateStatus(new()
+    {
+        dto.Id = Guid.Empty;
+
+        if (dto.Address is not null)
```
why do we need all these checks if we set default values ?

HlibBondarev replied:
We get the WorkshopCreateUpdateDto object from Frontend. It may contain not only null value but also invalid values, so we will have to replace them with the default value in the SetIdsToDefaultValue() method.
The Address property  of WorkshopBaseDto may be empty, so I need to check the Address before setting the Id property in Address object.
The same situation with the DefaultTeacher and MemberOfWorkshop properties.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -1165,4 +1140,77 @@ await UpdateStatus(new()
+            dto.MemberOfWorkshop.Id = Guid.Empty;
+        }
+
+        dto.WorkshopDescriptionItems?.ToList().ForEach(e => e.Id = Guid.Empty);
```
so in a worse case we will have a list of a billion items with default ID ? Why ?

HlibBondarev replied:
We will have as many items as will be transferred to DTO from the frontend and we need to set all of them value - Guid.Empty. Then DB will be generate new Id for WorkshopDescriptionItems. If we don't set Guid.Empty DB will use passed through Dto Id value. This will throw an exception if we pass the same Id value multiple times.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -1165,4 +1140,77 @@ await UpdateStatus(new()
+        return createdWorkshop;
+    }
+
+    private async Task SetIdsToDefaultValue(WorkshopCreateUpdateDto dto)
```
is there at least one good reason why we cannot create a new object instead of that ?

HlibBondarev replied:
We create a new Workshop from WorkshopCreateUpdateDto which we get from frontend. It may have contain invalid Id value in passed Dto for properties - Address, DefaultTeacher, MemberOfWorkshop, WorkshopDescriptionItems, Teachers, etc...

h4wk13 commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V2/WorkshopController.cs:
```diff
@@ -146,18 +161,46 @@ public async Task<IActionResult> GetByFilter([FromQuery] WorkshopFilter filter,
+            return StatusCode(403, "Forbidden to create the workshop by the blocked provider.");
+        }
+
+        if (!ModelState.IsValid)
```
should not this check be performed before checks for blocked users ?

HlibBondarev replied:
I assumed that check - 'is user blocked?' would take less time than validation of Model. That's why I chose this order. Is this wrong?

h4wk13 commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V2/WorkshopController.cs:
```diff
@@ -146,18 +161,46 @@ public async Task<IActionResult> GetByFilter([FromQuery] WorkshopFilter filter,
+            // here we will get "false" if workshop was created by assistant provider admin
+            // because user is not currently associated with new workshop
+            // so we can update information to allow assistant manage created workshop
+            if (!await IsUserProvidersOwnerOrAdmin(creationResult.Workshop.ProviderId, creationResult.Workshop.Id).ConfigureAwait(false))
```
it does not look like a controller logic, as for me it is more business logic and should be moved to the service

HlibBondarev replied:
I didn't do any WorkshopController logic (for V2), I just used the logic from WorkshopController (for V1) and put it in the Tray-Catch block.

h4wk13 commented on OutOfSchool/OutOfSchool.WebApi/Startup.cs:
```diff
@@ -191,7 +192,12 @@ public static async Task AddApplicationServices(this WebApplicationBuilder build
-            .AddNewtonsoftJson()
+
+            // to handle self-references during serialization
+            .AddNewtonsoftJson(options =>
```
are you sure it won't be an issue ?
I mean, we should find the root cause but not ignore the issue

HlibBondarev replied:
I need to handle self-references during serialization for WorkshopCreateUpdateDto because of Workshop entity has self-references.   
Workshop can have Workshops inside itself, also Workshop can be part of another Workshop. For these relationships are used MemberOfWorkshop and IncludedStudyGroups properties.

h4wk13 commented on OutOfSchool/OutOfSchool.WebApi/Startup.cs:
```diff
@@ -191,7 +192,12 @@ public static async Task AddApplicationServices(this WebApplicationBuilder build
             })
-            .AddNewtonsoftJson()
+
+            // to handle self-references during serialization
```
This comment explain 'what' this code is doing which is bad, but does not explain 'why' this code it is here.

HlibBondarev replied:
See above

HlibBondarev committed:
1) Replaced 'uint.MinValue' with '0' for uint properties of WorkshopBaseDto class.
2) Used tagRepository.GetByFilterNoTracking(tag => dto.TagIds.Contains(tag.Id)) method instead of (await tagRepository.GetAll()).Where(tag => dto.TagIds.Contains(tag.Id)) in WorkshopServices.

HlibBondarev committed:
Removed IsBlocked property from Workshop class.

HlibBondarev committed:
1) Changed CheckDtoAndPrepareCreatedWorkshop() method of WorkshopService class:
Replaced -
createdWorkshop.Tags = (await tagRepository.GetAll()).Where(tag => dto.TagIds.Contains(tag.Id)).ToList();
With -
createdWorkshop.Tags = (await tagRepository.GetByFilter(tag => dto.TagIds.Contains(tag.Id))).ToList();
2) Changed WorkshopConfiguration class - deleted relations between Tags and Workshop.
3) Deleted migration - 20241030104634_CreateWorkshop.

HlibBondarev committed:
1) Replaced 'uint.MinValue' with '0' for uint properties of Workshop class.
2) Added migration - 20241122215448_CreateWorkshop.

HlibBondarev committed:
1) Create request Dto classes WorkshopCreateRequestDto and WorkshopV2CreateRequestDto for creation Workshop.
2) Changed IWorkshopService and WorkshopService - used WorkshopCreateRequestDto instead of WorkshopCreateUpdateDto and WorkshopV2CreateRequestDto instead of WorkshopV2Dto.
3) Changed IWorkshopServicesCombiner and WorkshopServicesCombiner - used WorkshopCreateRequestDto instead of WorkshopCreateUpdateDto.
4) Changed IWorkshopServicesCombinerV2 and WorkshopServicesCombinerV2 - used WorkshopV2CreateRequestDto instead of WorkshopV2Dto.
5) Changed MappingProfile class - added mapping WorkshopCreateRequestDto -> Workshop and WorkshopV2CreateRequestDto -> Workshop.
6) Changed the argument type for the Create() method of WorkshopController(V1) - WorkshopCreateRequestDto instead of WorkshopCreateUpdateDto.
7) Changed the argument type for the Create() method of WorkshopController(V2) - WorkshopV2CreateRequestDto instead of WorkshopV2Dto.

HlibBondarev committed:
Fixed SetIdsToDefaultValue() method of WorkshopService class - throw InvalidOperationException when the DefaultTeacherId property of WorkshopBaseDto is incorrect.

HlibBondarev committed:
Removed redundant comments from classes - WorkshopCreateRequestDto, WorkshopService and MappingProfile.

HlibBondarev committed:
1) Changed MappingProfile class:
- fixed mapping WorkshopCreateRequestDto -> Workshop and WorkshopV2CreateRequestDto -> Workshop;
- added mapping Workshop -> WorkshopCreateRequestDto and Workshop -> WorkshopV2CreateRequestDto for tests.
2) Changed CheckDtoAndPrepareCreatedWorkshop() method of WorkshopService class - used workshopRepository.GetById() instead of WorkshopService.GetById().
3) Fixed tests in classes -  WorkshopServiceTests, WorkshopServicesCombinerTests, WorkshopServicesCombinerV2Tests, WorkshopControllerTests and WorkshopControllerV2Tests.

HlibBondarev committed:
Removed redundant comments from WorkshopCreateRequestDto class.

HlibBondarev committed:
Added migration - 20241122215448_CreateWorkshop.

HlibBondarev committed:
20241127184406_CreateWorkshop
